mlflow/tracing/export/mlflow.py

```diff
@@ -70,24 +76,159 @@ def export(self, root_spans: Sequence[ReadableSpan]):
+
+    def put(self, task: Task) -> None:
+        """Put a new task to the queue for processing."""
+        if not self.is_active():
+            self.activate()
+
+        # If stop even is set, we should wait for the queue to be drained before putting the task.
+        if self._stop_event.is_set():
+            self._stop_event.wait()
+
+        self._unprocessed_tasks.add(task)
+        self._queue.put(task)
+
+    def _logging_loop(self) -> None:
+        """
+        Continuously process incoming tasks in the queue until the stop event is set.
+        When the stop event is set, the loop will drain the remaining tasks in the queue.
+        """
+        while not self._stop_event.is_set():
+            self._handle_task()
+        while not self._queue.empty() or self._unprocessed_tasks:
+            self._handle_task()
+
+    def _handle_task(self) -> None:
+        """Process the given task in the running runs queues."""
+        try:
+            task = self._queue.get(timeout=1)
+        except Empty:
+            return
+
+        def _handle(task):
+            task.handle()
+            self._unprocessed_tasks.remove(task)
+
+        self._trace_logging_worker_threadpool.submit(_handle, task)
+
+    def activate(self) -> None:
+        """Activates the async logging queue"""
+        with self._lock:
+            if self._is_activated:
+                return
+
+            self._set_up_logging_thread()
+            # Registering an atexit callback to ensure that any remaining log data
+            # is flushed before the program exits.
+            atexit.register(self._at_exit_callback)
+            self._is_activated = True
+
+    def is_active(self) -> bool:
+        return self._is_activated
+
+    def _set_up_logging_thread(self) -> None:
+        """Sets up the logging thread."""
+        with self._lock:
+            self._trace_logging_thread = threading.Thread(
+                target=self._logging_loop,
+                name="MLflowAsyncTraceLoggingLoop",
+                daemon=True,
+            )
+            self._trace_logging_worker_threadpool = ThreadPoolExecutor(
+                max_workers=5,
+                thread_name_prefix="MLflowTraceLoggingWorkerPool",
+            )
+            self._trace_logging_thread.start()
+
+    def _at_exit_callback(self) -> None:
+        """Callback function to be executed when the program is exiting."""
+        try:
+            self.flush(terminate=True)
+        except Exception as e:
+            _logger.error(f"Encountered error while trying to finish logging: {e}")
+
+    def flush(self, terminate=False) -> None:
+        """
+        Flush the async logging queue.
+
+        Args:
+            terminate: If True, shut down the logging threads after flushing.
+        """
+        if not self.is_active():
+            return
+
+        self._stop_event.set()
+        self._trace_logging_thread.join()
+        self._trace_logging_worker_threadpool.shutdown(wait=True)
+
```

Comment:
Do we want to clear the unprocessed tasks as well, or is the intention of the flush to only reset the aync handler system and resume processing of still-in-flight elements?

Reply:
They are flushed in current implementation, basically the logging loop only stops when it exhaust the queue **and** unprocessed tasks.
```
        while not self._queue.empty() or self._unprocessed_tasks:
            self._handle_task()
```